File: src/Leitor/LeitorPeriodo.java
package Leitor;

import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.io.Serializable;
import java.util.HashMap;

import Modelo.Periodo;

public class LeitorPeriodo extends ILeitor implements Serializable{
    
    private static final long serialVersionUID = 1L;
    private HashMap<String, Periodo> mapa = new HashMap<>();
    private static LeitorPeriodo leitor;

    private LeitorPeriodo(){}

    public static LeitorPeriodo obterInstancia(){
        if (leitor == null){
            leitor = new LeitorPeriodo();
            return leitor;
        }
        else return leitor;
    }

    public HashMap<String,Periodo> obterHash(){
        return mapa;
    }

    public static void setInstancia(LeitorPeriodo novo){
        leitor = novo;
    }

    public Periodo ler(Scanner scanner) throws Exception {
        int ano = 0;
        
        try {
            ano = scanner.nextInt();
        } catch (Exception e) {
            throw new Exception("\nDado inválido: "+ scanner.next()+".\n");
        }

        String semestre = "";
        semestre = scanner.next();
        Pattern pattern = Pattern.compile("[0-9A-Za-z]");
        Matcher matcher = pattern.matcher(semestre);
        if(!matcher.matches())
            throw new Exception("Dado inválido: "+ semestre +".");

        Periodo periodo = new Periodo(ano, semestre.charAt(0));

        if(busca(periodo.obterCodigo()) != null)
            throw new IllegalArgumentException("Cadastro repetido: "+periodo.obterCodigo()+".");

        return periodo;
    }

    public void anexaHash(Periodo periodo){
        mapa.put(periodo.obterCodigo(), periodo);
    }

    public Periodo busca(String codigo){
        if(mapa.containsKey(codigo)){
            return mapa.get(codigo);
        }
        return null;
    }
}

File: cpp_version/Leitor/LeitorPeriodo.cpp
#include "LeitorPeriodo.h"

LeitorPeriodo* LeitorPeriodo::leitor = NULL;

LeitorPeriodo::LeitorPeriodo(){}

LeitorPeriodo* LeitorPeriodo::obterInstancia(){
    if (leitor == NULL){
        leitor = new LeitorPeriodo();
        return leitor;
    }
    else return leitor;
}

vector<Periodo*> LeitorPeriodo::obterPeriodos(){
    vector<Periodo*> lista;
    for(map<string,Periodo*>::iterator it = mapPeriodo.begin(); it != mapPeriodo.end(); ++it){
        lista.push_back(it->second);
    }
    return lista;
}

Periodo* LeitorPeriodo::ler(vector<string> vec){
    int ano;
    try{
        ano = stoi(vec[0]);
    } catch (...){
        throw ExcecaoCad("Dado inválido: "+ vec[0] +".");
    }

    string semestre = vec[1];

    regex rgx("^[0-9A-Za-z]$");
    smatch match;
    if(!regex_search(semestre, match, rgx))
        throw ExcecaoDado("Dado inválido: "+semestre+".");

    Periodo* periodo = new Periodo(ano, semestre[0]);

    if(busca(periodo->obterCodigo()) != NULL)
        throw ExcecaoCad("Cadastro repetido: "+periodo->obterCodigo()+".");

    return periodo;
}

void LeitorPeriodo::anexaHash(Periodo* periodo){
    mapPeriodo[periodo->obterCodigo()] = periodo;
}

Periodo* LeitorPeriodo::busca(string codigo){
    bool contains = false;
    for(map<string,Periodo*>::iterator it = mapPeriodo.begin(); it != mapPeriodo.end(); ++it){
        if(it->first == codigo){
            contains = true;
        }
    }
    if(contains){
        return mapPeriodo[codigo];
    }
    return NULL;
}